perf/src/sigverify.rs

```diff
@@ -420,34 +423,76 @@ pub fn generate_offsets(
+        let seed = thread_rng().gen();
+        Self {
+            filter,
+            seed,
+            age: Instant::now(),
+            max_age: Duration::from_millis(max_age_ms),
+            saturated: AtomicBool::new(false),
+        }
     }
 
-    // If this packet was not newly added, it's a dup and should be discarded
-    if !bloom.add(&&packet.data.as_slice()[0..packet.meta.size]) {
-        packet.meta.set_discard(true);
-        count.fetch_add(1, Ordering::Relaxed);
+    pub fn reset(&mut self) {
+        let now = Instant::now();
+        //this should reset every 500k unique packets per 1m sized deduper
+        //false positive rate is 1/1000 at that point
+        let saturated = self.saturated.load(Ordering::Relaxed);
+        if saturated || now.duration_since(self.age) > self.max_age {
+            for i in &self.filter {
+                i.store(0, Ordering::Relaxed);
+            }
+            self.seed = thread_rng().gen();
+            self.age = now;
+            self.saturated.store(false, Ordering::Relaxed);
+        }
     }
-}
 
-pub fn dedup_packets(bloom: &AtomicBloom<&[u8]>, batches: &mut [PacketBatch]) -> u64 {
-    use rayon::prelude::*;
-    let packet_count = count_packets_in_batches(batches);
-    // machine specific random offset to read the u64 from the packet signature
-    let count = AtomicU64::new(0);
-    PAR_THREAD_POOL.install(|| {
-        batches.into_par_iter().for_each(|batch| {
+    fn dedup_packet(&self, count: &AtomicU64, packet: &mut Packet) {
+        // If this packet was already marked as discard, drop it
+        if packet.meta.discard() {
+            return;
+        }
+        let mut hasher = AHasher::new_with_keys(self.seed.0, self.seed.1);
+        hasher.write(&packet.data[0..packet.meta.size]);
+        let hash = hasher.finish();
+        let len = self.filter.len();
+        let pos = (usize::try_from(hash).unwrap()).wrapping_rem(len);
+        // saturate each position with or
+        let prev = self.filter[pos].fetch_or(hash, Ordering::Relaxed);
+        if prev == u64::MAX {
+            self.saturated.store(true, Ordering::Relaxed);
+            //reset this value
+            self.filter[pos].store(hash, Ordering::Relaxed);
+        }
+        if hash == prev & hash {
+            packet.meta.set_discard(true);
+            count.fetch_add(1, Ordering::Relaxed);
+        }
+    }
+
+    pub fn dedup_packets(&self, batches: &mut [PacketBatch]) -> u64 {
+        let count = AtomicU64::new(0);
+        batches.iter_mut().for_each(|batch| {
             batch
                 .packets
-                .par_iter_mut()
-                .for_each(|p| dedup_packet(&count, p, bloom))
-        })
-    });
-    inc_new_counter_debug!("dedup_packets_total", packet_count);
-    count.load(Ordering::Relaxed)
+                .iter_mut()
+                .for_each(|p| self.dedup_packet(&count, p))
+        });
+        count.load(Ordering::Relaxed)
```

Comment:
As an alternative to passing in the atomic count to `dedup_packet()`, that fn could return a count of dupes, which could be summed up here. That would remove shared atomics, which may help if there's a lot of concurrent access of stuff on this same cache line.

i.e. here would be:

```rust
pub fn dedup_packets(&self, batches: &mut [PacketBatch]) -> u64 {
        batches.iter_mut().map(|batch| {
            batch
                .packets
                .iter_mut()
                .map(|p| self.dedup_packet(&count, p))
                .sum::<u64>()
        })
        .sum()
}
```

And `dedup_packet()` would return a u64. Or, it could return an `Option<()>`, too probably.